mltensor-explainer.md

```diff
@@ -0,0 +1,426 @@
+
+The machine learning context underlying WebNN may require input and output tensors to be allocated in a specific fashion, such as with a given byte alignment or on a given compute unit (e.g. CPU, GPU, NPU, TPU, etc...). Currently, this requires that the implementation of an `MLGraph` copy in data from the input tensors, execute the graph, and then copy out data from the output tensors.
+
+An `MLTensor` is an opaque tensor which may be created, written to, and read from independently from `MLGraph` inference. Each of these operations is performed on the [timeline](\#timelines) of the associated MLContext, with a clearly defined order of operations. Passing `MLTensor`s as input and output tensors to `MLGraph` inference - as opposed to passing `ArrayBufferView`s as is done today - allows for a decoupling of the uploading/downloading of model inputs/outputs from the model execution itself. This provides several benefits, such as buffer reuse, chained inference, explicit destruction, and the opportunity to interop with WebGPU.
+
+## Goals
+
+- Provide a consistent API for passing tensors to an `MLGraph` which may run arbitrary compute units (e.g. CPU, GPU, NPU, TPU, etc...)
+- Improve model throughput by minimizing the need for synchronization of work via JavaScript
+- Minimize overall data copies of graph inputs and outputs
+- Best-effort buffer-sharing between WebNN and WebGPU
+- Allow a tensor to be reused across multiple `MLGraph`s within the same `MLContext`
+
+## Non-Goals
+
+* Guarantee *zero-copy* buffer-sharing between WebNN and WebGPU
+* Synchronization of work between WebNN and WebGPU without CPU involvement
+* Provide partial views over an `MLTensor`
+
+## Key Scenarios
+
+### Buffer Reuse
+
+A user uploads a large image to a website and wants to try applying several different effects on the same input image.
+
+The current way that WebNN passes input buffers requires a copy to be made _for each inference_.
+
+```js
+// Current approach to reuse a given input buffer, requiring many data copies
+
+// Copy the data in `imageAsArrayBuffer` to the required format before executing `graph1`.
+// Then, copy the model outputs into `outputArrayBuffer1`.
+await mlContext.compute(graph1, {'input': imageAsArrayBuffer}, {'output': outputArrayBuffer1});
+// Again, copy all input data in and all output data out.
+await mlContext.compute(graph2, {'input': imageAsArrayBuffer}, {'output': outputArrayBuffer2});
+// Yet again copy all input data in and all output data out.
+await mlContext.compute(graph3, {'input': imageAsArrayBuffer}, {'output': outputArrayBuffer3});
+```
+
+`MLTensor` allows tensors and their contents to be reused. Once the image data is written to an `MLTensor`, no further copies are required.
+
+```js
+// Proposed approach to reuse a given input buffer, using an input MLTensor
+
+// Copy the image data into the required format.
+const imageAsMlTensor = await mlContext.createTensor({..., usage: MLTensorUsage.WRITE});
```

Comment:
You might want to use normal enums (strings) instead of constants, per https://w3ctag.github.io/design-principles/#string-constants

Reply:
Oh yeah, that looks a bit nicer than nested dictionaries to me. I don't think there's any guidance regarding them but my intuition would be to only have nested dictionaries if the inner one is some sort of shared type that is used across multiple APIs.

Reply:
Agreed. Cool, the latest patch includes the suggested diff